clang/test/SemaSYCL/kernelname-enum.cpp

```diff
@@ -0,0 +1,79 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -fsycl-int-header=%t.h -fsyntax-only -verify %s
+
+//#include <sycl.hpp>
+
+#include "sycl.hpp"
+
+enum unscoped_enum_int : int {
+  val_1,
+  val_2
+};
+
+// expected-note@+1 {{'unscoped_enum_no_type_set' declared here}}
+enum unscoped_enum_no_type_set {
+  val_3,
+  val_4
+};
+
+enum class scoped_enum_int : int {
+  val_1,
+  val_2
+};
+
+enum class scoped_enum_no_type_set {
+  val_3,
+  val_4
+};
+
+template <unscoped_enum_int EnumType>
+class dummy_functor_1 {
+public:
+  void operator()() {}
+};
+
+// expected-error@+2 {{kernel name is invalid. Unscoped enum requires fixed underlying type}}
+template <unscoped_enum_no_type_set EnumType>
```

Comment:
It would be great if this diagnostic will also point to concrete `single_task`/`parallel_for` kernel invocation call which caused this error. See https://github.com/intel/llvm/blob/9afcfad942485f61088329cbeafd65ff81ef006d/clang/test/SemaSYCL/unnamed-kernel.cpp#L51 for example.

Reply:
This is because kernel location is set differently when kernel is defined using a lambda vs a functor. I just tested kernelname-enum.cpp using lambdas instead of functors and the diagnostic is generated at the point of kernel invocation like this example above. Since this is an existing issue with all diagnostics emitted when kernel is defined using functors, I'd like to work on this as a separate issue if that is alright with you.

Reply:
Okay. We need to capture this problem somehow.

Reply:
Ok. I'll file a bug and assign it to me.

Reply:
https://github.com/intel/llvm/issues/1706